quickwit/quickwit-ingest/src/ingest_v2/workbench.rs

```diff
@@ -17,22 +17,108 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
-use std::collections::{BTreeMap, HashSet};
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::sync::{Arc, Mutex};
 
+use quickwit_common::pubsub::{EventBroker, EventSubscriptionHandle};
 use quickwit_common::rate_limited_error;
 use quickwit_proto::control_plane::{
     GetOrCreateOpenShardsFailure, GetOrCreateOpenShardsFailureReason,
 };
+use quickwit_proto::indexing::ShardPositionsUpdate;
 use quickwit_proto::ingest::ingester::{PersistFailure, PersistFailureReason, PersistSuccess};
 use quickwit_proto::ingest::router::{
     IngestFailure, IngestFailureReason, IngestResponseV2, IngestSubrequest, IngestSuccess,
 };
 use quickwit_proto::ingest::{IngestV2Error, RateLimitingCause};
-use quickwit_proto::types::{NodeId, ShardId, SubrequestId};
+use quickwit_proto::types::{NodeId, Position, ShardId, SubrequestId};
+use tokio::sync::Notify;
 use tracing::warn;
 
 use super::router::PersistRequestSummary;
 
+#[derive(Default)]
+struct PublishState {
+    awaiting_publish: HashMap<ShardId, Position>,
+    already_published: HashMap<ShardId, Position>,
```

Comment:
could we create the publish state with the list of ShardId we have an interest in?
The point would be to avoid cloning and saving all of the shards (thousands) getting published in Quickwit.

Reply:
I hesitated doing so, but I wasn't quite sure that maintaining that extra filter would be cheaper than recording all publish events. If the published shard event bus can be very busy, it's definitively worth it.

Reply:
I think we want to support at least 10k indexes, and commit every 30s.

Reply:
changed the state machine to track only requested shards